x-pack/plugins/fleet/server/services/agent_policy.ts

```diff
@@ -531,6 +533,166 @@ class AgentPolicyService {
     };
   }
 
+  public async bulkUpdate(
+    soClient: SavedObjectsClientContract,
+    esClient: ElasticsearchClient,
+    agentPolicies: Array<Partial<AgentPolicy> & { id: string }>,
+    options: {
+      user?: AuthenticatedUser;
+      force?: boolean;
+      skipValidation?: boolean;
+      spaceId?: string;
+      authorizationHeader?: HTTPAuthorizationHeader | null;
+      removeProtection?: boolean;
+      bumpRevision?: boolean;
+    }
+  ): Promise<{
+    updatedPolicies: AgentPolicy[] | null;
+    failedPolicies: Array<{ agentPolicy: Partial<AgentPolicy>; error: Error | SavedObjectError }>;
+  }> {
+    const logger = appContextService.getLogger();
+    logger.info(`Starting bulk update of agent policies: ${agentPolicies.map(({ id }) => id)}`);
+
+    const existingAgentPolicies = await this.list(soClient, {
+      perPage: SO_SEARCH_LIMIT,
+      withPackagePolicies: true,
+    });
+
+    this.checkNameUniquenessBulk(agentPolicies, existingAgentPolicies.items);
+    this.checkPolicyExistsBulk(agentPolicies, existingAgentPolicies.items);
+
+    const policiesToUpdate: Array<Partial<AgentPolicy> & { id: string }> = [];
+    const failedPolicies: Array<{
+      agentPolicy: Partial<AgentPolicy>;
+      error: Error | SavedObjectError;
+    }> = [];
+
+    await pMap(agentPolicies, async (agentPolicy) => {
+      try {
+        this.checkTamperProtectionLicense(agentPolicy);
+        await this.checkForValidUninstallToken(agentPolicy, agentPolicy.id);
+        const existingAgentPolicy = existingAgentPolicies.items.find(
+          (i) => i.id === agentPolicy.id
+        );
+        if (!existingAgentPolicy) {
+          throw new AgentPolicyNotFoundError('Agent policy not found'); // This should never happen
```

Comment:
Not very helpful of an error here - can you add the agent policy ID that was not found?